common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3986,31 +3986,25 @@ static jobject GENERAL_NAME_to_jobject(JNIEnv* env, GENERAL_NAME* gen) {
             /* Write in RFC 2253 format */
             return X509_NAME_to_jstring(env, gen->d.directoryName, XN_FLAG_RFC2253);
         case GEN_IPADD: {
-#ifdef _WIN32
-            void* ip = reinterpret_cast<void*>(gen->d.ip->data);
-#else
-            const void* ip = reinterpret_cast<const void*>(gen->d.ip->data);
-#endif
-            if (gen->d.ip->length == 4) {
-                // IPv4
-                std::unique_ptr<char[]> buffer(new char[INET_ADDRSTRLEN]);
-                if (inet_ntop(AF_INET, ip, buffer.get(), INET_ADDRSTRLEN) != nullptr) {
-                    JNI_TRACE("GENERAL_NAME_to_jobject(%p) => IPv4 %s", gen, buffer.get());
-                    return env->NewStringUTF(buffer.get());
-                } else {
-                    JNI_TRACE("GENERAL_NAME_to_jobject(%p) => IPv4 failed %s", gen,
-                              strerror(errno));
-                }
-            } else if (gen->d.ip->length == 16) {
-                // IPv6
-                std::unique_ptr<char[]> buffer(new char[INET6_ADDRSTRLEN]);
-                if (inet_ntop(AF_INET6, ip, buffer.get(), INET6_ADDRSTRLEN) != nullptr) {
-                    JNI_TRACE("GENERAL_NAME_to_jobject(%p) => IPv6 %s", gen, buffer.get());
-                    return env->NewStringUTF(buffer.get());
-                } else {
-                    JNI_TRACE("GENERAL_NAME_to_jobject(%p) => IPv6 failed %s", gen,
-                              strerror(errno));
-                }
+            const uint8_t* data = ASN1_STRING_get0_data(gen->d.ip);
+            if (ASN1_STRING_length(gen->d.ip) == 4) {
+                // IPv4 addresses are returned in dotted quad notiation.
+                char buffer[4 * 3 + 3 + 1];  // 4 3-digit fields, 3 separators, and a NUL
+                snprintf(buffer, sizeof(buffer), "%d.%d.%d.%d", data[0], data[1], data[2], data[3]);
+                JNI_TRACE("GENERAL_NAME_to_jobject(%p) => IPv4 %s", gen, buffer);
+                return env->NewStringUTF(buffer);
+            } else if (ASN1_STRING_length(gen->d.ip) == 16) {
+                // IPv6 addresses are returned as eight 16-bit components. Note this does not quite
+                // match inet_ntop, which will abbreviate "0:0:0:0:0:0:0:1" as "::1".
```

Comment:
That's fine though, as the documentation explicitly states it will be returned in the long format

Reply:
Oh, that's what I meant. I think I just worded the comment confusingly. Is this new wording clearer? (I just meant this comment to explain why we we're _not_ using inet_ntop.)

Reply:
Oh right, gotcha..Thanks!